keps/sig-api-machinery/3285-subresource-finalizers/README.md

```diff
@@ -0,0 +1,812 @@
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Risks:
+- Bugs could make unintended fields accessible.
+
+Mitigations:
+- The new endpoint needs to be negative-tested to prove that the other fields are NOT modifiable.
+
+## Design Details
+
+Add `<resource>/finalizers` endpoint to all resources and allow access to `metadata.finalizers` for the resources through the endpoints.
+The actual field in each resource isn't added, but the request to the endpoint is handled as the operation to `metadata.finalizers`.
+`Role` and `ClusterRole` can be set to the endpoint, therefore users can set a separate RBAC to `<resource>/finalizers`.
+A PoC implementation can be found [here](https://github.com/mkimuram/kubernetes/commit/751a67e03d62397c1d21947e9e2e3a81f279a29f) (Please also see [here](https://github.com/kubernetes/enhancements/pull/2840#discussion_r723464012) for how it works. Note that API definition in this KEP will differ from the one in the PoC).
+
+```
+<<[UNRESOLVED Endpoint name]>>
+Endpoint name for the subresource is still being discussed.
+- `/finalize`: already used by namespace
+- `/finalizers`: some debate about whether sub-resources should be nouns of verbs
+<<[/UNRESOLVED]>>
+```
+
+### API Changes
+
+A new `Metadata` resource is introduced in `subresources.k8s.io` API group:
+
+```golang
+// Metadata represents a metadata field to be accessed via an endpoint for the subresource.
+type Metadata struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Only following fields should be set:
+	// - Namespace
+	// - Name
+	// - UID
+	// - ResourceVersion
+	// - The field for the endpoint, like Finalizers field for the finalizers endpoint
+	// If other fields are set, they are ignored.
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+}
+```
+
+```
+<<[UNRESOLVED Resource name]>>
+Resource name is still being discussed.
+`Metadata` seems good if the same type can be used for all of the metadata subresources.
+If resource should be defined for each metadata subresource, `Finalize` or `Finalizers` will be candidates.
+<<[/UNRESOLVED]>>
+```
+
+### Supported API calls to the endpoint
+
+The `<resource>/finalizers` endpoint supports 2 verbs, GET and PUT.
+
+```
+<<[UNRESOLVED Whether to implement PATCH]>>
+There is no good way to remove a specific item from a list by using existing PATCH request.
+Server-side apply is trying to resolve this, but it can't still be used for this purpose, currently.
+Therefore, whether to implement PATCH is decided later.
+<<[/UNRESOLVED]>>
+```
+
+#### GET
+
+GET verb is used to read the value in `metadata.finalizers` field.
+
+- Request format:
+    Empty string should be passed as a request body. Any strings passed in the request body are ignored.
+
+- Response format:
+  - On success, the `Metadata` resource is returned. `Namespace`, `Name`, `UID`, `ResourceVersion`, and `Finalizers` fields are copied from the original resources.
```

Comment:
Why wouldn't we include all metadata?

Reply:
I am biased to less-is-more.  Why would we have a GET permission per subresource if any one of them gets all of them?

Reply:
Because we have a GET-PUT correspondence baked into the system.